finat/direct_serendipity.py

```diff
@@ -0,0 +1,371 @@
+        # assert degree == 1 or degree == 2
+        self._cell = cell
+        self._degree = degree
+        self.space_dim = 4 if degree == 1 else (self.degree+1)*(self.degree+2)//2 + 2
+
+    @property
+    def cell(self):
+        return self._cell
+
+    @property
+    def degree(self):
+        return self._degree
+
+    @property
+    def formdegree(self):
+        return 0
+
+    def entity_dofs(self):
+        if self.degree == 1:
+            return {0: {i: [i] for i in range(4)},
+                    1: {i: [] for i in range(4)},
+                    2: {0: []}}
+        elif self.degree == 2:
+            return {0: {i: [i] for i in range(4)},
+                    1: {i: [i+4] for i in range(4)},
+                    2: {0: []}}
+        else:
+            return {0: {i: [i] for i in range(4)},
+                    1: {i: list(range(4 + i * (self.degree-1),
+                                      4 + (i + 1) * (self.degree-1)))
+                        for i in range(4)},
+                    2: {0: list(range(4 + 4 * (self.degree - 1),
+                                      self.space_dim))}}
+
+    def space_dimension(self):
+        return self.space_dim
+
+    @property
+    def index_shape(self):
+        return (self.space_dimension(),)
+
+    @property
+    def value_shape(self):
+        return ()
+
+    def basis_evaluation(self, order, ps, entity=None, coordinate_mapping=None):
+        '''Return code for evaluating the element at known points on the
+        reference element.
+
+        :param order: return derivatives up to this order.
+        :param ps: the point set.
+        :param entity: the cell entity on which to tabulate.
+        '''
+        ct = self.cell.topology
+
+        # Build everything in sympy
+        if self.degree == 1:
+            vs, xx, phis = ds1_sympy(ct)
+        else:
+            vs, xx, phis = dsr_sympy(ct, self.degree)
+
+        # and convert -- all this can be used for each derivative!
+        phys_verts = coordinate_mapping.physical_vertices()
+
+        phys_points = gem.partial_indexed(
+            coordinate_mapping.physical_points(ps, entity=entity),
+            ps.indices)
+
+        repl = {vs[i, j]: phys_verts[i, j] for i in range(4) for j in range(2)}
+
+        repl.update({s: phys_points[i] for i, s in enumerate(xx)})
+
+        mapper = gem.node.Memoizer(sympy2gem)
+        mapper.bindings = repl
+
+        result = {}
+        for i in range(order+1):
+            alphas = mis(2, i)
+            for alpha in alphas:
+                dphis = [phi.diff(*tuple(zip(xx, alpha))) for phi in phis]
+                result[alpha] = gem.ListTensor(list(map(mapper, dphis)))
+
+        return result
+
+    def point_evaluation(self, order, refcoords, entity=None):
+        raise NotImplementedError("Not done yet, sorry!")
+
+    def mapping(self):
+        return "physical"
+
+
+def xysub(x, y):
+    return {x[0]: y[0], x[1]: y[1]}
+
+
+def ds1_sympy(ct):
```

Comment:
Maybe some brief comments describing which equations in the paper you're implementing?

Reply:
Also docstring, especially describing both the input argument, and the return value.